runtime-light/stdlib/file/file-system-functions.cpp

```diff
@@ -0,0 +1,314 @@
+        if (cur_bits > bits) {
+          bits = cur_bits;
+        }
+      }
+      bits++;
+    } else {
+      return false;
+    }
+    break;
+  case 0x00: // jp2
+    if (read_size >= 54 && !std::strncmp((const char*)buf.begin(), php_sig_jp2.begin(), sizeof(php_sig_jp2))) {
+      type = IMAGETYPE_JP2;
+
+      bool found = false;
+
+      int buf_pos = 12;
+      size_t file_pos = 12;
+      while (static_cast<int>(read_size) >= 42 + buf_pos + 8) {
+        const unsigned char* s = buf.begin() + buf_pos;
+        int box_length = (s[0] << 24) + (s[1] << 16) + (s[2] << 8) + s[3];
+        if (box_length == 1 || box_length > 1000000000) {
+          break;
+        }
+        if (s[4] == 'j' && s[5] == 'p' && s[6] == '2' && s[7] == 'c') {
+          s += 8;
+
+          width = (s[8] << 24) + (s[9] << 16) + (s[10] << 8) + s[11];
+          height = (s[12] << 24) + (s[13] << 16) + (s[14] << 8) + s[15];
+          channels = (s[40] << 8) + s[41];
+
+          if (channels < 0 || channels > 256 || static_cast<int>(read_size) < 42 + buf_pos + 8 + 3 * channels || width <= 0 || height <= 0) {
+            break;
+          }
+
+          bits = 0;
+          for (int i = 0; i < channels; i++) {
+            int cur_bits = s[42 + 3 * i];
+            if (cur_bits > bits) {
+              bits = cur_bits;
+            }
+          }
+          bits++;
+
+          found = true;
+          break;
+        }
+
+        if (box_length <= 8) {
+          break;
+        }
+        file_pos += box_length;
+        if (file_pos >= size || static_cast<off_t>(file_pos) != static_cast<ssize_t>(file_pos) || static_cast<ssize_t>(file_pos) < 0) {
+          break;
+        }
+
+        read_size = min_size;
+        if (size - file_pos < min_size) {
+          read_size = size - file_pos;
+        }
+
+        if (read_size < 50) {
+          break;
+        }
+        std::span<unsigned char> buf_read_span(buf.begin(), read_size);
+        read_res = f->pread(std::as_writable_bytes(buf_read_span), static_cast<off_t>(file_pos));
+        if (!read_res || *read_res < read_size) {
+          break;
+        }
+
+        buf_pos = 0;
+      }
+
+      if (!found) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+    break;
+  case 0x89: // png
+    if (read_size >= 25 && !std::strncmp((const char*)buf.begin(), php_sig_png.begin(), sizeof(php_sig_png))) {
+      type = IMAGETYPE_PNG;
+      width = (buf[16] << 24) + (buf[17] << 16) + (buf[18] << 8) + buf[19];
+      height = (buf[20] << 24) + (buf[21] << 16) + (buf[22] << 8) + buf[23];
+      bits = buf[24];
+    } else {
+      return false;
+    }
+    break;
+  default:
+    return false;
+  }
+
+  array<mixed> result(array_size(7, false));
+  result.push_back(width);
+  result.push_back(height);
+  result.push_back(type);
+
+  string::size_type len = std::format_to_n(reinterpret_cast<char*>(buf.data()), min_size, "width=\"%d\" height=\"%d\"", width, height).size;
```

Comment:
format_to_n returns not truncated output size. To calculate the recorded size let's use std::distance(buf.begin(), result.out)